### PR TITLE
Feature/spp 95 historytab back periods

### PR DESCRIPTION
**Pull request by sivagudur**
Business Layer changes for getting History Data

### CODE REVIEW

File: src/main/java/uk/gov/ons/collection/controller/ViewFormController.java [hunk 1]
```diff
@@ -40,6 +44,45 @@ public String viewFormDetails(@MatrixVariable Map<String, String> searchParamete
...
+            List<String> historyPeriodList = responsePeriodicity.getHistoryPeriods(currentPeriod, periodicityStr);
+            log.info("Final History Periods: " + historyPeriodList.toString());
+            if (historyPeriodList.size() > 0) {
+                String historyQuery = detailsQuery.buildHistoryDetailsQuery(historyPeriodList);
+                log.info("History Details Query: " + historyQuery);
+                String historyDetailsResponse = qlService.qlSearch(historyQuery);
+                log.info("History Details Response: " + historyDetailsResponse);
+                responseText = responsePeriodicity.parseHistoryDataResponses(historyDetailsResponse);
+            }
+
```
Comment (dllmorgan) [inline src/main/java/uk/gov/ons/collection/controller/ViewFormController.java:76]: Multiple calls here which can generate exceptions. Should we have more than 1 try--catch here so we can be more specific about which bit has failed?

File: src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        return historyPeriodList;
+    }
+
+    public String parseHistoryDataResponses(String responseJson) throws InvalidIdbrPeriodException {
+        JSONObject queryOutput = new JSONObject(responseJson);
+        JSONArray contribArray = new JSONArray();
+        var historyDataObj = new JSONObject();
+        var historyDataArr = new JSONArray();
+        try {
+            if (queryOutput.getJSONObject("data").getJSONObject("allContributors").getJSONArray("nodes").length() > 0) {
```
Comment (dllmorgan) [inline src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java:81]: Branch condition == 0 is not covered by the unit tests

File: src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                contribArray = queryOutput.getJSONObject("data").getJSONObject("allContributors").getJSONArray("nodes");
+                for (int i = 0; i < contribArray.length(); i++) {
+                    JSONObject eachContributorObj = queryOutput.getJSONObject("data").getJSONObject("allContributors").getJSONArray("nodes").getJSONObject(i);
+                    var historyDetailsObject = new JSONObject();
+                    historyDetailsObject.put(REFERENCE, eachContributorObj.get(REFERENCE));
+                    historyDetailsObject.put(PERIOD, eachContributorObj.get(PERIOD));
+                    historyDetailsObject.put(SURVEY,eachContributorObj.get(SURVEY));
+                    JSONArray formDefinitionArray = eachContributorObj .getJSONObject("formByFormid")
+                            .getJSONObject("formdefinitionsByFormid").getJSONArray("nodes");
+                    if (formDefinitionArray.length() > 0) {
```
Comment (dllmorgan) [inline src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java:91]: Branch condition == 0 is not covered by the unit tests

**Review by dllmorgan — COMMENTED** [review]
Only a quick eyeball review.

File: src/main/java/uk/gov/ons/collection/controller/ViewFormController.java (hunk 1 above)
Comment (sivagudur) [reply src/main/java/uk/gov/ons/collection/controller/ViewFormController.java]: Done. I added 2 try catch exceptions, one for periodicity and another for getting history data.

File: src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java (hunk 2 above)
Comment (sivagudur) [reply src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java:82]: I added a unit test and also throwing an exception in development code which specifically says "There are no contributors for a given survey, reference and periods. Please verify"

File: src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java (hunk 3 above)
Comment (sivagudur) [reply src/main/java/uk/gov/ons/collection/entity/HistoryDetailsResponse.java:92]: Done. I added a unit test and also throwing an exception in development code which specifically says "There is no form definition associated to a given Contributor. Please verify"
Also executed Jacoco code coverage and both HistoryDetailsResponse.java & HistoryDetailsQuery covers 100% code coverage.

**Review by mattprice-28 — APPROVED** [review]
quick overview - unit tests passing locally and expected behaviour observed in the UI

**Review by dllmorgan — APPROVED** [review]

**Pull request merged**